### ISSUE TITLE
clickhouse 20.6.1.1 restart error: external dictionary 'rtype' not found: Cannot attach table...

### ISSUE THREAD

**Issue by MrSsunlight** [ClickHouse/ClickHouse#12454]
version: 20.6.1.1
system: centos7

**description：**
-  No exception is reported at the first startup, the dictionary is loaded successfully；
-  No other operations were done during the restart, only the service was restarted, but the dictionary loading exception was reported after restarting clickhouse；

```
Logging errors to /var/log/clickhouse-server/clickhouse-server.err.log
Logging error to console
2020.07.13 15:38:24.756706 [ 641 ] {} <Error> Application: Caught exception while loading metadata: Code: 36, e.displayText() = DB::Exception: external dictionary 'rtypes' not found: Cannot attach table `dnsmon`.`rtypes` from metadata file /var/lib/clickhouse/metadata/dnsmon/rtypes.sql from query ATTACH TABLE rtypes (`value` UInt16, `rtype` String) ENGINE = Dictionary('rtypes'), Stack trace (when copying this message, always include the lines below):

0. /work/ClickHouse/contrib/poco/Foundation/src/Exception.cpp:27: Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0xc65e91c in /usr/bin/clickhouse
1. /work/ClickHouse/src/Common/Exception.cpp:37: DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x54c8a19 in /usr/bin/clickhouse
2. /work/ClickHouse/contrib/libcxx/include/string:2134: DB::ExternalLoader::checkLoaded(DB::ExternalLoader::LoadResult const&, bool) const (.cold) @ 0x50c0b4b in /usr/bin/clickhouse
3. /work/ClickHouse/src/Interpreters/ExternalLoader.h:65: std::__1::shared_ptr<DB::IExternalLoadable const> DB::ExternalLoader::load<std::__1::shared_ptr<DB::IExternalLoadable const>, void>(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) const @ 0x975911c in /usr/bin/clickhouse
4. /work/ClickHouse/contrib/libcxx/include/memory:4047: std::__1::__function::__func<DB::registerStorageDictionary(DB::StorageFactory&)::'lambda'(DB::StorageFactory::Arguments const&), std::__1::allocator<DB::registerStorageDictionary(DB::StorageFactory&)::'lambda'(DB::StorageFactory::Arguments const&)>, std::__1::shared_ptr<DB::IStorage> (DB::StorageFactory::Arguments const&)>::operator()(DB::StorageFactory::Arguments const&) @ 0x9a7f04c in /usr/bin/clickhouse
5. /work/ClickHouse/contrib/libcxx/include/string:2134: DB::StorageFactory::get(DB::ASTCreateQuery const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::Context&, DB::Context&, DB::ColumnsDescription const&, DB::ConstraintsDescription const&, bool) const @ 0x9a8e117 in /usr/bin/clickhouse
6. /work/ClickHouse/contrib/libcxx/include/utility:447: DB::createTableFromAST(DB::ASTCreateQuery, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::Context&, bool) @ 0xa05b07f in /usr/bin/clickhouse
7. /work/ClickHouse/contrib/libcxx/include/memory:4081: DB::(anonymous namespace)::tryAttachTable(DB::Context&, DB::ASTCreateQuery const&, DB::DatabaseOrdinary&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, bool) @ 0xa0655c8 in /usr/bin/clickhouse
8. /work/ClickHouse/contrib/libcxx/include/string:2134: std::__1::__function::__func<DB::DatabaseOrdinary::loadStoredObjects(DB::Context&, bool)::'lambda0'(), std::__1::allocator<DB::DatabaseOrdinary::loadStoredObjects(DB::Context&, bool)::'lambda0'()>, void ()>::operator()() @ 0xa065f1e in /usr/bin/clickhouse
9. /work/ClickHouse/contrib/libcxx/include/atomic:856: ThreadPoolImpl<ThreadFromGlobalPool>::worker(std::__1::__list_iterator<ThreadFromGlobalPool, void*>) @ 0x54d2fdd in /usr/bin/clickhouse
10. /work/ClickHouse/src/Common/ThreadPool.h:167: ThreadFromGlobalPool::ThreadFromGlobalPool<void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()>(void&&, void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()&&...)::'lambda'()::operator()() const @ 0x54d36c3 in /usr/bin/clickhouse
11. /work/ClickHouse/contrib/libcxx/include/atomic:856: ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0x54d257d in /usr/bin/clickhouse
12. /work/ClickHouse/contrib/libcxx/include/memory:2615: void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()> >(void*) @ 0x54d0c4f in /usr/bin/clickhouse
13. start_thread @ 0x7ea5 in /usr/lib64/libpthread-2.17.so
14. __clone @ 0xfe8dd in /usr/lib64/libc-2.17.so
 (version 20.6.1.1)
2020.07.13 15:38:25.733448 [ 641 ] {} <Error> Application: DB::Exception: external dictionary 'rtypes' not found: Cannot attach table `dnsmon`.`rtypes` from metadata file /var/lib/clickhouse/metadata/dnsmon/rtypes.sql from query ATTACH TABLE rtypes (`value` UInt16, `rtype` String) ENGINE = Dictionary('rtypes')
```

**The corresponding configuration details are as follows**

```
<?xml version="1.0"?>
<yandex>
    <dictionary>
        <name>rtypes</name>
        <source>
            <file>
		    <path>/etc/clickhouse-server/config.d/rtype.csv</path>
                <format>TSV</format>
            </file>
        </source>
        <lifetime>300</lifetime>
        <layout>
            <complex_key_hashed/>
        </layout>
        <structure>
            <key>
                <attribute>
                    <name>value</name>
                    <type>UInt16</type>
                    <null_value>0</null_value>
                </attribute>
            </key>
            <attribute>
               <name>rtype</name>
               <type>String</type>
               <null_value />
            </attribute>
        </structure>
    </dictionary>
</yandex>
```

> [root@826f2ed32868 /]# ll /etc/clickhouse-server/config.d/rtype.csv 
> -rw-r--r-- 1 root root 719 Jul 10 18:49 /etc/clickhouse-server/config.d/rtype.csv

> [root@826f2ed32868 /]# cat /var/lib/clickhouse/metadata/dnsmon/rtypes.sql
```
ATTACH TABLE rtypes
(
    `value` UInt16,
    `rtype` String
)
ENGINE = Dictionary('rtypes')
```

I look at related questions, there are still some differences, such as [https://github.com/ClickHouse/ClickHouse/issues/10994]， But did not get a clear solution。

**Comment by MrSsunlight** [comment]
If I move `/var/lib/clickhouse/metadata/dnsmon/rtypes.sql /tmp/` before starting clickhouse, clickhouse can start successfully；And, after the startup is successful, move the file back to the source path, attach the table in clickhouse-client, it can be successfully loaded, no exception information

**Comment by den-crane** [comment]
duplicate https://github.com/ClickHouse/ClickHouse/issues/10994

It's broken in 20.4 and 20.5 you can use 20.3.

it has been fixed in testing 20.6.1.4050+

**Comment by MrSsunlight** [comment]
> duplicate #10994
> 
> It's broken in 20.4 and 20.5 you can use 20.3.
> 
> it has been fixed in testing 20.6.1.4050+

thanks

**Comment by asafcombo** [comment]
reproduced on 20.12.5 revision 54442

**Comment by alexey-milovidov** [comment]
@asafcombo How exactly does it reproduce? Could you please create a new issue?

**Comment by asafcombo** [comment]
new issue at : 

[21101](https://github.com/ClickHouse/ClickHouse/issues/21101)
